### PR TITLE
feat(EthereumToOptimism): add dispatchAndProcess function

### PR DESCRIPTION
**Pull request by PierrickGT**
Changes:
- update dependencies
- fix lint errors and remove Optimism lib
- add dispatchAndProcess functions

### CODE REVIEW

**Comment by PierrickGT** [comment]
Arbitrum fork tests are not working because we need to update Hardhat dependencies but when doing so, files fail to compile because Hardhat does not support ESM for TypeScript projects.
Here is the related issue: https://github.com/NomicFoundation/hardhat/issues/3385

File: src/ethereum-optimism/EthereumToOptimismDispatcher.sol [hunk 1]
```diff
@@ -151,12 +166,79 @@ contract MessageDispatcherOptimism is ISingleMessageDispatcher, IBatchedMessageD
...
+   * @dev Must emit the `MessageBatchDispatched` event when successfully dispatched.
+   * @param _toChainId ID of the receiving chain
+   * @param _messages Array of Message dispatched
+   * @param _gasLimit Gas limit at which the messages will be executed on Optimism
+   * @return bytes32 ID uniquely identifying the `messages`
+   */
+  function _dispatchMessageBatch(
+    uint256 _toChainId,
+    MessageLib.Message[] calldata _messages,
+    uint32 _gasLimit
```
Comment (trmid) [inline src/ethereum-optimism/EthereumToOptimismDispatcher.sol:218]: Same here: Is this the gas limit for all messages together, or does each one individually gets this gas limit?

File: src/ethereum-optimism/EthereumToOptimismDispatcher.sol [hunk 2]
```diff
@@ -56,41 +57,51 @@ contract MessageDispatcherOptimism is ISingleMessageDispatcher, IBatchedMessageD
...
+   * @dev Must emit the `MessageBatchDispatched` event when successfully dispatched.
+   * @param _toChainId ID of the receiving chain
+   * @param _messages Array of Message dispatched
+   * @param _gasLimit Gas limit at which the message will be executed on Optimism
+   * @return bytes32 ID uniquely identifying the `messages`
+   */
+  function dispatchMessageWithGasLimitBatch(
+    uint256 _toChainId,
+    MessageLib.Message[] calldata _messages,
+    uint32 _gasLimit
```
Comment (trmid) [inline src/ethereum-optimism/EthereumToOptimismDispatcher.sol:102]: Is this the gas limit for all messages together, or does each one individually gets this gas limit?

Comment (PierrickGT) [reply src/ethereum-optimism/EthereumToOptimismDispatcher.sol:102]: It's for all messages together.

File: src/ethereum-optimism/EthereumToOptimismDispatcher.sol (hunk 1 above)
Comment (PierrickGT) [reply src/ethereum-optimism/EthereumToOptimismDispatcher.sol:218]: It's for all messages together.
We pass the gas limit to the bridge: https://github.com/GenerationSoftware/ERC5164/blob/f8690d9a3304f3469e855ac5486c544e89f5a5c7/src/ethereum-optimism/EthereumToOptimismDispatcher.sol#L244
And then all messages are being executed one by one: https://github.com/GenerationSoftware/ERC5164/blob/f8690d9a3304f3469e855ac5486c544e89f5a5c7/src/libraries/MessageLib.sol#L177
If not enough gas, the whole batch reverse.

**Review by trmid — APPROVED** [review]

**Pull request merged**